File: desktop/tests/Qlipq.Core.Tests/EditSpecTests.cs
using Qlipq.Core;
using Xunit;

namespace Qlipq.Core.Tests;

public class EditSpecTests
{
    private static readonly MediaInfo Media = new()
    {
        DurationSec = 120,
        Width = 1920,
        Height = 1080,
        VideoCodec = "h264",
        Fps = 60,
        AudioStreams =
        [
            new AudioStreamInfo { StreamIndex = 1, Index = 0, Codec = "aac", Channels = 2, Title = "Desktop" },
            new AudioStreamInfo { StreamIndex = 2, Index = 1, Codec = "aac", Channels = 1, Title = "Mic" },
        ],
    };

    [Fact]
    public void DefaultEditSpecEnablesEverySourceTrackAtUnityGain()
    {
        var spec = EditSpecs.DefaultEditSpec(Media);
        Assert.Equal(
            [
                new AudioTrackSpec { Index = 0, Enabled = true, Volume = 1 },
                new AudioTrackSpec { Index = 1, Enabled = true, Volume = 1 },
            ],
            spec.AudioTracks);
        Assert.Null(spec.Trim);
    }

    [Fact]
    public void EffectiveDurationReflectsTrimWindow()
    {
        Assert.Equal(120, EditSpecs.EffectiveDuration(new EditSpec { AudioTracks = [] }, Media));
        Assert.Equal(30, EditSpecs.EffectiveDuration(
            new EditSpec { AudioTracks = [], Trim = new TrimSpec { StartSec = 10, EndSec = 40 } }, Media));
    }

    [Fact]
    public void ValidateEditSpecAcceptsASaneSpec()
    {
        var spec = EditSpecs.DefaultEditSpec(Media);
        spec.Trim = new TrimSpec { StartSec = 5, EndSec = 50 };
        spec.Crop = new CropSpec { X = 0, Y = 0, Width = 1280, Height = 720 };
        Assert.Null(EditSpecs.ValidateEditSpec(spec, Media));
    }

    [Fact]
    public void ValidateEditSpecRejectsInvertedTrim()
    {
        var msg = EditSpecs.ValidateEditSpec(
            new EditSpec { AudioTracks = [], Trim = new TrimSpec { StartSec = 30, EndSec = 10 } }, Media);
        Assert.Contains("after the start", msg);
    }

    [Fact]
    public void ValidateEditSpecRejectsCropOutsideFrame()
    {
        var msg = EditSpecs.ValidateEditSpec(
            new EditSpec { AudioTracks = [], Crop = new CropSpec { X = 1000, Y = 0, Width = 1280, Height = 720 } }, Media);
        Assert.Contains("outside the frame", msg);
    }

    [Fact]
    public void ValidateEditSpecRejectsNegativeVolume()
    {
        var msg = EditSpecs.ValidateEditSpec(
            new EditSpec { AudioTracks = [new AudioTrackSpec { Index = 0, Enabled = true, Volume = -1 }] }, Media);
        Assert.Contains("volume cannot be negative", msg);
    }
}

File: iced/crates/qlipq-core/tests/edit_spec_tests.rs
use qlipq_core::edit_spec::*;
use qlipq_core::media::{AudioStreamInfo, MediaInfo};

fn media() -> MediaInfo {
    MediaInfo {
        duration_sec: 120.0,
        width: 1920,
        height: 1080,
        video_codec: "h264".into(),
        fps: 60.0,
        audio_streams: vec![
            AudioStreamInfo { stream_index: 1, index: 0, codec: "aac".into(), channels: 2, language: None, title: Some("Desktop".into()) },
            AudioStreamInfo { stream_index: 2, index: 1, codec: "aac".into(), channels: 1, language: None, title: Some("Mic".into()) },
        ],
        size_bytes: None,
    }
}

#[test]
fn default_edit_spec_enables_every_track_at_unity() {
    let spec = default_edit_spec(Some(&media()));
    assert_eq!(
        spec.audio_tracks,
        vec![
            AudioTrackSpec { index: 0, enabled: true, volume: 1.0 },
            AudioTrackSpec { index: 1, enabled: true, volume: 1.0 },
        ]
    );
    assert!(spec.trim.is_none());
}

#[test]
fn effective_duration_reflects_trim() {
    let m = media();
    assert_eq!(effective_duration(&EditSpec { trim: None, crop: None, audio_tracks: vec![] }, &m), 120.0);
    assert_eq!(
        effective_duration(
            &EditSpec { trim: Some(TrimSpec { start_sec: 10.0, end_sec: 40.0 }), crop: None, audio_tracks: vec![] },
            &m
        ),
        30.0
    );
}

#[test]
fn validate_accepts_sane_spec() {
    let mut spec = default_edit_spec(Some(&media()));
    spec.trim = Some(TrimSpec { start_sec: 5.0, end_sec: 50.0 });
    spec.crop = Some(CropSpec { x: 0, y: 0, width: 1280, height: 720 });
    assert_eq!(validate_edit_spec(&spec, &media()), None);
}

#[test]
fn validate_rejects_inverted_trim() {
    let spec = EditSpec { trim: Some(TrimSpec { start_sec: 30.0, end_sec: 10.0 }), crop: None, audio_tracks: vec![] };
    assert!(validate_edit_spec(&spec, &media()).unwrap().contains("after the start"));
}

#[test]
fn validate_rejects_crop_outside_frame() {
    let spec = EditSpec { trim: None, crop: Some(CropSpec { x: 1000, y: 0, width: 1280, height: 720 }), audio_tracks: vec![] };
    assert!(validate_edit_spec(&spec, &media()).unwrap().contains("outside the frame"));
}

#[test]
fn validate_rejects_negative_volume() {
    let spec = EditSpec { trim: None, crop: None, audio_tracks: vec![AudioTrackSpec { index: 0, enabled: true, volume: -1.0 }] };
    assert!(validate_edit_spec(&spec, &media()).unwrap().contains("volume cannot be negative"));
}
